[1단계 - 장기] 투다(박찬양) 미션 제출합니다.

praisebak opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
🦆안녕하세요 오리! 이번에 리뷰를 받게된 투다입니다
부족한 코드이지만 잘부탁드리겠습니다!
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/4104

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
평소에 설계에 시간을 많이 들이지 않았는데, 페어가 설계에 시간을 한번 많이 두는것이 어떻겠냐해서 한번 시도해보았습니다.
이후 직접 경험해보니 개인적인 의견은, 시간을 많이 들이고 공을 들여도 수정을 안할 수 있는 것은 아니라고 생각되었고
설계에 시간 들인것에 비해 결과물이 아쉬워서 4점을 주었습니다. 
또한 설계에 시간을 많이 들이니 구현에 신경을 적게써서 버그가 있었고 설계당시에는 신경쓰지 못한, 놓친 요구사항이 많았습니다.


이번에 경험을 통해 다음에는 아마 요구사항을 이해했다면 TDD로 진행해서 설계에대한 고민은 이후 리팩토링단계에서 진행할것같습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?
Piece,Pieces,움직임 방식에 대해서 설계에 많은 시간을 들였습니다. (설계기반으로 한번 해보고 싶다는 페어의 의견을 적용했습니다)
- 그중 책임의 명확함,캡슐화에 집중하였습니다.
- 설계에 너무 많은 시간을 쓴 나머지 구현과 테스트, 리팩토링에는 시간을 거의 쏟지 못한것같습니다.
- 좀 더 구현과 테스트로 일단 잘 돌아가는 프로그램을 완성하고, 그 뒤에 리팩토링을 해도 괜찮았지않나 하는 아쉬움이 남았습니다 

- 그래서 버그가 많을것으로 예상됩니다.
- 코드가 더럽고 통일되지않은 부분이 많습니다! 시간에 쫒혀 구현하면서 무의식적으로 습관적으로 작성한 코드가 많을것같은데 이런 무의식적으로 잘못작성했던 코드들을 알고싶어요. 

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

### 다음은 페어를 진행하며 논의가 있었던 내용들 입니다

#### 기물이 스스로 좌표를 가지는 설계에서 움직임을 직접 처리하게하려니 외부에서 메시지를 여러번 줘야함
- 이렇게 했을때 기물의 구현은 조금 복잡해지지만 기물의 책임이 명확해지는 장점이 있다
- 기물이 단지 VO가되고 외부에서 이를 처리하면 외부 객체의 책임이 더 커질것이다
- 구현이 복잡한것보다는 객체지향적인 가치인 응집도, 캡슐화가 더 중요하다 판단하였습니다.
>- pieces와 협력해서 가는길에 있는 기물을 찾아서 piece에게 넘겨준다(”piece 너 가는길에 이거있는데 갈수있으면 가”)
- pieces에서해버리면 너무 pieces의 책임이 커진다 - 관리도해야하고 움직임도 관리해야하고..등


### 기물 초기 설정 파일 VS Enum
- 외부에서 주입받는 방식 vs enum으로 주입받는방식

#### enum 장점)
- enum 방식의 초안을 작성해보니, 테스트 용의성면에서 enum들을 가지고 처리할 수 있어서 편했습니다.
그런데 이것은 파일방식도 테스트 용이하려면 만들 수 있을것같았습니다.
-  enum 방식의 경우에는 장기판 초기화에만 사용되는 부분인데, 코드로 존재하는것이 아쉬웠습니다.

#### 파일 방식 장점
- 파일로 만들면 초기화만 잘 되는것이 테스트되면 코드는 더 신경쓰지않아도 된다는것이 장점이라고 생각하여 적용해보았습니다.
- 파일방식의 단점은 파싱이 제대로 되어야한다는것을 보장해줘야하는것이 단점인 것 같았습니다.
- 외부에 의존되는것은 초기에 1회만 사용되는것이니 상관없다고 생각하였습니다.

### 많은 구현체 클래스 문제
- 현재 기물 움직임 클래스가 기물 움직임만큼 존재하고 있습니다.
- 런타임으로 기물들을 만드는 방식을 생각해보았으나, config에서 담을 설정이 복잡하고 이를 파싱한다면 너무 복잡할것 같았습니다.
- 기물들의 동작과 특성이 다르고, 사용되는 생명주기가 장기만큼 동일하게때문에 런타임방식보다 명확하게 구현체를 구현하는 방식을 택하였습니다.
 

### 팩토리,맵퍼 도입
- 팩토리, 맵퍼 객체가 필요한가
적합한 움직임 Strategy를 뽑아낼 역할을 수행할 객체가 존재하지않아 팩토리를 도입해보았습니다.

### 중복된 관리 문제
- 컬렉션을 사용하는 객체들이 관리해야하는 대상이 겹치는 문제가 계속 발생하였습니다.
- 예로 (현재 코드와는 다른 구조입니다)Board 클래스가 위치로  Map<Position, Piece>을 관리한다고했을때, Piece에서도 Position을 가지는 상황에서,
Piece의 Position이 불변으로 바뀐다고할때, Board의 필드의 Position도 업데이트해줘야하는 문제가 발생하였습니다
따라서 관리 대상이 겹치는 것을 지양하였습니다. 물론 지양할 수 있으면 좋지만, 구현하다보니 겹치는 경우도 생겼습니다

예로 Board(현재 코드와 같은 구조입니다) 에서 Map<Team,Pieces>으로 팀단위로 피스들을 관리, 청팀의 Pieces에서는 List<Piece>로 기물을 제거하는경우를 가정하면,
청팀 Pieces에서 제거 - Board의 Pieces도 변경되어서 사이드이펙트로 예상하기 어려운 코드가 됩니다.
그런데 실질적으로 이것을 아예 피하는 방법은 아직 생각하지 못했습니다.

관리 객체가 겹쳐서 **예상하기 어려운 코드**가 발생하는것, 지양하면 좋겠지만 100% 지켜야만하는것일지 오리의 의견이 궁금합니다!
- 위 문제를 중복된 관리가 없어도 되는 방향으로 리팩토링해보고 싶습니다!

### 성능문제
- Map 대신 List - 어차피 장기말은 30개밖에 되지 않으니 전체 탐색이 상관없다. 오히려 메모리가 더 이득이다!

Hacanna42 committed:
feat: Position 클래스 추가

Hacanna42 committed:
feat: PiecesCreateFactory 클래스 추가

Hacanna42 committed:
feat: Pieces 클래스 추가

praisebak committed:
feat: 피스들 생성기 구현

praisebak committed:
fix: 컴파일 에러 해결

Hacanna42 committed:
feat(JolMoveStrategy): 졸의 이동 가능 경로 반환 구현

praisebak committed:
test: pieces 테스트 추가

praisebak committed:
test: 졸 가는길에 기물 없을시 이동 가능 테스트 추가

Hacanna42 committed:
test: 졸 이동 테스트 추가

Hacanna42 committed:
feat: 차 이동 기능 추가

praisebak committed:
test: 사,궁 테스트 추가

praisebak committed:
feat: 포 움직임 구현

praisebak committed:
feat: 마 움직임 구현

praisebak committed:
feat: 마 움직임 구현

praisebak committed:
feat: 상 움직임 구현

Hacanna42 committed:
feat: 기물 잡기 기능 추가

Hacanna42 committed:
docs: 기능 명세 추가

Hacanna42 committed:
chore: 구조 추가

praisebak committed:
feat: 유저 체스 기능 추가

praisebak committed:
fix: piece의 remove 기능과 Board 계층 연결

praisebak committed:
refactor: 컨벤션 준수

praisebak committed:
refactor: 런타임에 변경되지않음 - 전략패턴이아닌 단지 행위이므로 명명 수정

praisebak commented on src/main/java/strategy/ChaMoveStrategy.java:
```diff
@@ -0,0 +1,36 @@
+            return calculateLegalRoute(minPosition, maxPosition, positions, new Position(0, 1));
+        }
+        throw new IllegalArgumentException(INVALID_MOVE_LOCATION);
+    }
```
개선하고 싶은 부분)
ChaMoveStrategy와

FoMoveStrategy의 중복코드가 발생하는 부분을 개선하고싶은데, SameLineMoveStrategy라는 이름으로 전략을 추가하여
SameLineMoveStrategy에서 default 메서드를 사용하여 없애면 더 깔끔해질것같은데,

이 경우 단점은 인터페이스 하나만보고 구현하는것은 힘들어질것같고, 의존성이 좀 복잡한것같아요.

그런데 이것이 정말 객체지향에서 고려해야할 단점인지 잘 모르겠어요.
즉 코드가 이해하기 어려운것은 객체지향의 트레이드오프의 저울위에 올라가면 안되지 않을까하는 생각입니다.

중복을 줄이는것은 클린코드에 가까운 코드를 만들수있는데, 코드가 이해하기 어려운것은 큰 가치에 축하는것은 아니지않은가 하는 생각입니다!

jinyoungchoi95 replied:
1. default 메서드를 활용하는 것은 객체지향인가요?
2. 마냥 중복을 줄이는 것이 클린코드인가요? 클린코드는 중복이 있으면 클린코드라 할 수 없나요?

praisebak replied:
> 마냥 중복을 줄이는 것이 클린코드인가요? 클린코드는 중복이 있으면 클린코드라 할 수 없나요?
## 가졌던 잘못된 생각
중복코드 = 클린코드를 위해서 무조건 지켜야하는 규칙이라고 생각했었어요.
## 생각해보니..
중복코드를 만들지말라는 내용은 유지보수가 힘든 것등의 이유가 있는데, 그 이유에 대해서 생각하지않고 원칙을 무조건적으로 지킬려고 했던것같습니다.

그러면 해당 코드에서 적용해봤을때는 차의 기물 이동 방식이 바뀐다고 가정하면, 포는 안바뀌어도 되기때문에 
굳이 이것을 중복코드여서 없애야한다는 결론은 내지 않아도 될것같습니다!

---
> default 메서드를 활용하는 것은 객체지향인가요?
- 상황에 따라 다를것같습니다. 완전히 isA라고 보는 설계에서는 중복코드를 줄일 수 있는 장점이 있습니다. (완전히 isA여서 중복코드 줄일수있고 안바뀌니까 상관없다고 보는 입장)

- 순수한 인터페이스의 측면에서는 인터페이스는 설계도여야한다는 측면에서 default 메서드를 오히려 지양해야할것같습니다.
  - isA가 깨지면 하위 클래스에서 모두 구현체를 만들어야하는데, default 메서드의 존재로 명확하게 이를 알려주지 않습니다.  

저는 인터페이스가 순수하게 설계도 일 필요는 없다고 생각하고 현재 움직임이라는 측면에서 isA라고 보고 있기 때문에 default 메서드를 사용해도 문제 없다는 입장으로 사용했습니다. 
- 기물들은 항상 외부에서부터 이동할려는 위치, 팀에 대한 정보를 받는것은 고정된다고 보고 
이때는 default 메서드를 잘 활용하면 중복코드만 줄일 수 있다고 생각하여 적용하였습니다

jinyoungchoi95 replied:
관련 답변은 아래 다른 default 코멘트에 달아두었어요.

praisebak committed:
refactor: 불필요한 주석 제거

praisebak commented on src/main/java/GameView.java:
```diff
@@ -0,0 +1,98 @@
+
+    public void playerBoard(Map<Position, Piece> positionPieceMap) {
+        StringBuilder stringBuilder = new StringBuilder(GRID_HELPER);
+        for (int i = 9; i >= 0; i--) {
+            stringBuilder.append(i).append(" ");
+            for (int j = 0; j < 9; j++) {
+                String type = pieceType(positionPieceMap, i, j);
+                stringBuilder.append(type);
+            }
+            stringBuilder.append("\n");
+        }
+        System.out.println(stringBuilder.toString());
```
intent를 2로 유지한 코드입니다.
메서드로 분리를하려고하니 메서드에 사이드이펙트를 내포하게되고,
하려고하는 일이 잘 유추가되지 않는것같아 그대로 유지하였습니다.

praisebak commented on src/main/java/GameView.java:
```diff
@@ -0,0 +1,98 @@
+        return inputPosition(s);
+    }
+
+    public Position inputPiecePosition() {
```
String을 필드로 빼지 않은 이유는
View에 한해서, 무엇을 출력하는지 확인을하기가 편하기때문에 위와 같이 유지하였습니다

praisebak committed:
refactor: 매직변수제거 및 테스트용주석 제거

praisebak commented on src/main/java/strategy/GungMoveStrategy.java:
```diff
@@ -0,0 +1,21 @@
+import piece.Route;
+import piece.Team;
+
+public class GungMoveStrategy implements MoveStrategy {
+
+    private static final String NOT_IMPLEMENTED_FEATURE = "아직 구현되지 않은 기능입니다.";
+
+    @Override
+    public Route getLegalRoute(Position startPosition, Position endPosition, Team team) {
+        throw new IllegalStateException(NOT_IMPLEMENTED_FEATURE);
+    }
+
+    @Override
+    public Position move(Position destination, Pieces onRoutePieces, Team moveTeam) {
+        throw new IllegalStateException(NOT_IMPLEMENTED_FEATURE);
+    }
+}
```
궁이나 사를 사용하면 예외가 발생합니다.
체크메이트도 구현하지않았습니다

praisebak commented on src/main/java/piece/Board.java:
```diff
@@ -0,0 +1,57 @@
+        return bluePieces.add(redPieces);
+    }
+
+    public boolean isKingDead() {
+        return false;
+    }
```
아직 구현되지않은 기능을 잘 표시하고싶은데, 예외를 발생시키자니 별로인것같았고
마땅한 방법이 잘 떠오르지않았어요

jinyoungchoi95 replied:
왕이 움직일 수 없는 것이지 다른 말에 의해서 죽을 수 는 있는 것 아닌가요? 구현되지 않는 이유가 있을까요?

praisebak replied:
시간이 부족해서 요구사항에서 말한 궁성을 고려한 왕의 이동을 구현하지 않아도되는것을 마음대로 형편에 좋게 받아들였던것같습니다

왕을 구현하도록 수정하겠습니다!

praisebak commented on src/test/java/piece/BoardTest.java:
```diff
@@ -0,0 +1,9 @@
+import org.junit.jupiter.api.DisplayNameGeneration;
+import org.junit.jupiter.api.DisplayNameGenerator;
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+public class BoardTest {
+
+}
```
Board는 Team단위로 Pieces를 관리하는 책임을 가지고 있습니다.
테스트를 작성하려고하니 Board의 public 메서드중 playerBoard(플레이어에게 장기판 보여주기)는 테스트를 작성하지않을려고합니다.(view 기능이기때문에)

그런데 move의 경우에는, pieces에게 명령을 내리는 부분이고 일부분 pieces가 필요한 파라미터를 넘겨주는 역할을하는데,
이러면 테스트를 작성하는것이 좋을까요?
현재는 이미 pieces에서 move 동작자체는 커버가 됐다고 생각하여서 작성하지 않은 상황입니다

jinyoungchoi95 replied:
> 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다

`Board`가 `Team단위로 Pieces를 관리하는 책임`을 가지고 있는데 view의 기능도 하고 있으면 요구사항을 깨는 것 아닌가요?

`playerBoard()`는 테스트해야하는 도메인 로직 아닌가요?

커버가 되었다고 말씀하셨는데 아래 기능이 커버되었나요?
```java
    public void move(Team team, Position selectPiecePosition, Position selectPosition) {
        Pieces pieces = teamBoard.get(team);
        otherTeamPieces(team);
        Piece piece = pieces.move(selectPiecePosition, selectPosition, allPieces());
        Pieces otherPieces = otherTeamPieces(team);
        pieces.killPieceFrom(piece, otherPieces);

    }

    private Pieces otherTeamPieces(Team team) {
        if (team == Team.BLUE) {
            return teamBoard.get(Team.RED);
        }
        return teamBoard.get(Team.BLUE);
    }
```

praisebak replied:
## 기존 생각
- Board의 move는 Pieces들의 동작을 시키기만한다고 생각했어요.
그리고 이런 하위계층들을 조합하는 일은 테스트를 하지 않아도 된다고 생각했었습니다.

## 기존 생각이 잘못된 이유
- 아무리 조합하는것이어도 Board의 역할이 테스트가 되지 않는것, 애초에 조합하는것도 기능이자 역할인데 역할을 테스트 하지 않는다는것으로 볼 수 있습니다 
- 역할을 가진다면 테스트를 작성하는 방향으로 수정하겠습니다!
- 기존에는 테스트를 너무 등한시 했었어서, 이번 피드백에서는 테스트를 요구사항에 대한 책임을 가지는 객체들을 1대1로 맵핑이 되도록 수정해보려고 합니다.

## playerBoard()와 관련 Board 책임 명확하지 않은 문제
- Board는 팀 별로 기물을 관리한다는 메시지를 가지고 있는데, 이름이 Board여서 뭔가 포지션별로 기물들을 관리하는것과 헷갈릴 것 같아 이름을 TeamPieces로 변경하도록 하겠습니다
- playerBoard가 뷰 책임이라고 말씀드렸었는데 그 이유는 유저에게 포지션별로 기물들을 출력하는것은 뷰측에서 할일이라고 생각했었습니다.
- 애초에 playerBoard는 컨트롤러에서 가지고 있어야했을것 같습니다.

## 추가로 궁금한점
- 객체 개수와 1:1로 테스트 클래스가 만들어질때, 테스트 코드도 관리를 해줘야하는데 테스트 클래스가 많아지니 관리한 코드도 많아집니다.
- 이부분에 대한 문제는 현재 생각했을때는 객체 개수가 많아질것이 두려워 객체를 만들지않는것처럼 테스트도 각 객체마다 있으면 다다익선이아닌가하는 생각이 드는데

**혹시 객체와 테스트를 1:1로 대응시키는것을 지양해야할 상황이 있을까요?**

jinyoungchoi95 replied:
그런 상황은 따로 없는 것 같아요.

praisebak commented on src/main/java/strategy/ChaMoveStrategy.java:
```diff
@@ -0,0 +1,36 @@
+
+public class ChaMoveStrategy implements MoveStrategy {
+
+    private static final String INVALID_MOVE_LOCATION = "이동불가능한 위치입니다.";
```
Cha뿐만아닌 다른 MoveStrategy에서 INVALID_MOVE_LOCATION를 중복하여 가지고 있습니다.
이동불가능한 위치라는 메시지를 던질만한 객체가 애매하고,
Enum으로 만들기에는 한개의 메시지만 필요해서
현재 어떤식으로 중복코드를 줄일지 고민중인 상황인데,
오리와 함께 어떻게 접근하는게 좋을지 고민해봐도될까요?

jinyoungchoi95 replied:
이동불가능한 위치에 해당하는 예외를 만들어볼 수 있겠네요.

praisebak replied:
이럴때 예외를 활용하는것은 생각치 못했던것같습니다 알려주셔서 감사합니다!

praisebak replied:
예외를 추가하고 적용하다보니, 함수형 인터페이스를 사용해서 가독성을 향상시켜보았는데,
MoveBehavior가 자신만 사용하는 인터페이스를 내부에 가지고 있어도 괜찮을지,
어떻게보면 유틸리티인데 유틸리티 범위가 너무 작은것인지 염려가 되었습니다.

위 부분 오리와 함께 얘기해보고 싶습니다!

praisebak replied:
[refactor: 중복된 움직임 예외 처리 부분 가독성 향상](https://github.com/woowacourse/java-janggi/pull/42/commits/3ec1f4e7436fcbba91400b0c09360a841d931d18)

jinyoungchoi95 replied:
유틸이라는 것을 잊고 exception도 하나의 도메인이라고 생각하고 배치해보세요. `exception` 패키지는 현재 미션에서 굳이 필요없는 것 같습니다.

jinyoungchoi95 commented on src/main/java/resources/initialPieces.txt:
> 기물 초기 설정 파일 VS Enum

기물의 초기 위치는 도메인 정책아닌가요? 파일로 관리하면 도메인을 잘 관리하고 있는 것 맞을까요?

praisebak replied:
## 파일 방식 선택
저는 기물의 위치를 외부에서 관리하는 설정으로 관리하고싶었습니다
그래서 파일을 선택하였습니다
우선 장기판이 바뀌는 상황을 생각했었습니다(일반 장기판말고 조커 장기판이라던지..)
저는 외부에서 장기판을 쉽게 바꿀수있으면 좋은게 아닌가? 생각했었어요.
그런데 오히려 장기판을 쉽게 바꿀수있고 파일이 제대로 생겼는지 테스트가 안되고있어서 개발자가 실수로 한글자만 바꿔도 
제 장기는 잘 돌아가지 않게 되었습니다.

왜 이런 생각을 했을까 생각하면, 그냥 파일 방식을 적용하고 싶었던것같아요. 그리고 파일 방식을 적용할때의 장점을 억지로 끼워맞춘것같습니다

장점과 단점이 정확히 맞는지를 좀 더 검수 해야할것같아요

그래서 초기 기물 상태값을 Factory로 가지고 있고 그것을 반환하는 방식으로 수정해보았습니다.

praisebak replied:
[장기 초기화 파일방식에서 enum 방식으로 변경](https://github.com/woowacourse/java-janggi/pull/42/commits/d60ed0cf63ad5110ac2abbc4254983f84e3ae411)

praisebak replied:
[장기 초기화 파일방식에서 초기 기물을 상태로서 가지고 있게 수정](https://github.com/woowacourse/java-janggi/pull/42/commits/ac863122678219b4c6d70172e856d7b5913c9387)

jinyoungchoi95 replied:
이미 "장기"이기 때문에 다른 게임인 체스를 실행하는 것에 대해서는 고려하지 않아도 될 것 같아요.

jinyoungchoi95 commented on src/main/java/strategy/MoveStrategyFactory.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.Map;
+import piece.PieceType;
+
+public class MoveStrategyFactory {
```
> 적합한 움직임 Strategy를 뽑아낼 역할을 수행할 객체가 존재하지않아 팩토리를 도입해보았습니다.

팩토리라는 것이 필요한지 잘 모르겠습니다. 디자인 패턴은 잊고, `MoveStrategyFactory`는 객체가 맞나요? 다른 객체와 협력하지 않고 static성 기능만 제공하는 것이 객체지향인가요?

praisebak replied:
## MoveStrategyFactory 도입 논의
PiecesCreateFactory를 사용하기 시작하면서 MoveRule이라는 이름을 봤을때 MoveStrategy를 직접 만들어도 될까?라는 논의가 있었습니다.
생각했을때 MoveStrategyFactory가 있으면 누가 생성하는지 의견이 있었고 함께 동의를 하여서 MoveStrategyFactory를 사용했습니다.

그럼 팩토리는 상태가 없는 객체가 되는데 이것이 객체가 맞고 객체지향이 맞느냐라는 논의가 거론될 수 있을것같습니다.
- 저는 객체지향이 장점을 명확성이라고 보고 있습니다. 어떤 기능을 추가해야할때 누구한테 시키는게 좋을까 라는 것을 
객체의 이름만 보고 유추할 수 있는 명확성이 최대 장점이라고 생각합니다.
위 관점에서 봤을때 Piece, 기물이 스스로 자신의 움직임을 정의한다는것이 저는 Piece에서 움직임을 만드는게 책임이 애매하다고 보았습니다.
그래서 Factory를 두면 이름만 보았을때 유추가 가능하여 팩토리를 도입하였습니다
디자인패턴을 염두에두고 사용하지는 않았습니다


## 이 방식의 단점
- 도메인에 static 메서드를 사용하는것의 단점을 그대로 가집니다.
- 도메인의 static을 사용한다는것은 해당 도메인이 바뀌지않는다는것을 가정한다고 생각합니다(static 메서드는 확장하기 힘든것으로 알고 있습니다)
- 도메인은 변경가능해야하고 확장가능해야하기 때문에 위 생각을 지양해야합니다.

위 2가지를 조합하면,
Factory라는 이름은 그대로 두되, static 방식이아닌 외부에서 주입받는 방식을 사용해볼려고합니다.

---

JanggiTeamPiecesCreateFactory에서 피스들을 만드는 책임을 가지고 있어서 해당 팩토리에서만
이를 잘 관리 해주면 되어서 MoveStrategyFactory는 삭제하였습니다

praisebak replied:
[필요하지않은 기물 이동방식 팩토리 제거](https://github.com/woowacourse/java-janggi/pull/42/commits/c948963130ce18698ebf6d8a4612d8b7dd53bba6)

jinyoungchoi95 commented on src/main/java/piece/Board.java:
```diff
@@ -0,0 +1,57 @@
+
+public class Board {
+
+    private final Map<Team, Pieces> teamBoard;
```
> 관리 객체가 겹쳐서 예상하기 어려운 코드가 발생하는것, 지양하면 좋겠지만 100% 지켜야만하는것일지 오리의 의견이 궁금합니다!

이해를 하지 못했어요. `Board`는 `Pieces`를 관리하는 것이고 `Piece`는 `List<Piece>`를 관리하는 것인데 어디가 관리 대상이 겹치고 어떤 사이드 이펙트가 발생하나요?

praisebak replied:
제 설명 부족으로 잘못 전달드렸습니다!
다시 설명 드리겠습니다
현재 Board에서 홍,청 팀별로 Pieces를 관리하고 있는데, 홍이 청의 Pieces에서 기물을 잡아먹는경우를 가정하겠습니다.
이 경우 Board안의 홍팀의 Pieces의 Piece가 한개 줄어야합니다.
즉 Red에 해당하는 Pieces안에서 변경을하는데 Board에서도 바뀌어야한다는 것을 관리 객체가 겹친다고 표현하였습니다.

- 위부분 저는 Pieces에서 알아서 잘 바꾸면 Board에서는 상관없는것이 아니냐는 의견인데 다음의 의견을 듣게 되었습니다. 
- Board는 Map<Tema,Pieces>를 보고 있는데 Pieces의 내부 상태가 바뀌는것인데 관리하는 객체로써 내부 상태 관리도 신경써야하는것이아니냐는 의견있었습니다 즉 애초에 이 구조로 가면 안된다는 의견입니다
위 부분 현재 논의가 지속되고 있는도중 오리의 의견이 궁금했습니다!

praisebak replied:
수업과 크루들과의 토론을 통해 결국 관리가 겹치는 부분도 트레이드오프의 영역이라는 생각을 가지게 되었습니다.
질문 드린 상황에서는 Map을 사용하면 요구사항을 수행하기 위한 코드가 명확해집니다(팀별로 Pieces가져오기 등)
대신에 동기화를 해야할 필요성이 생깁니다.

위 두가지를 생각해보았을때 이번에는 동기화가 크게 어렵지 않고 명확해지는 부분은 크다 생각하여 현재 방식을 유지해보려 해요.

jinyoungchoi95 replied:
`Board`의 관리대상이 홍/청 `Pieces`모두이기 때문에 청이 움직여 홍이 제거되는 것이 이상해보이진 않아요.

명확하게 한다면 `Position`이 key이고 팀 정보를 가진 Piece가 value인 것이 실제 장기판을 떠올려보면 더 자연스럽겠네요.

jinyoungchoi95 commented on src/main/java/piece/Position.java:
```diff
@@ -0,0 +1,65 @@
+        return (this.row + this.column) - (position.row + position.column) > 0;
+    }
+
+    public static Position getMinPosition(Position position, Position otherPosition) {
+        if (position.isAbsoluteBigger(otherPosition)) {
+            return otherPosition;
+        }
+        return position;
+    }
+
```
static이 아닌 객체와 객체가 메시지를 주고받는 형태로 만들어보세요. static 메서드는 객체지향에 부합하나요?

praisebak replied:
위 코드를 처음에는 Position a, Position b
a.getMin(b);와 같은 식으로 구현하는것을 생각했었습니다
그런데 Position이 하는 getMin이 뭘 리턴하는건지 명확하지 않다는 의견이 있어서 한번 static 방식으로 구현해보았습니다

저는 당시에 static 메서드를 사용해서 유틸성임을 나타내는것은 괜찮다고 생각하였고 크게 static의 단점을 생각하지 않았었습니다
찾아보니 static 메서드는 다음의 문제점들이 있었습니다. 
상태 고려 X 다형성 X 테스트 어려움

다시 생각해보면, 위 코드가 명확하지않다는것이 메서드 명명이 문제지 static을 사용해서 해결되는 문제는 아니었던것같아 수정하려고 합니다.

praisebak replied:
[사용하지 않는 코드 제거 및 Position 불필요한 static 사용 제거](https://github.com/woowacourse/java-janggi/pull/42/commits/94231ee2b54b998b68211abdc56fe0528d4125cd)

jinyoungchoi95 commented on src/main/java/piece/Team.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Arrays;
+
+public enum Team {
+    RED("홍"),
+    BLUE("청"),
```
> 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.

화면의 요구사항이 영어로 바뀌면 화면의 요구사항으로 인해 도메인 로직이 변경되어야겠네요.

praisebak replied:
음 위 부분은 지금은 그대로 사용하고 있지만  요구사항이 영어로 바뀌면 뷰에서 조건문으로 처리할것같습니다!
위와같이 구현한다고 해도 Team의 도메인 로직이 변경되어야할까요?

혹시 Team의 from에서 String으로 Team을 찾아오는데 이것이 현재 파일에서 의존된 상태여서 문제가 생길수있다는 말씀일까요?

jinyoungchoi95 replied:
`Team.getType()`으로 출력문이 사용되고 있기 때문에 View의 요구사항이 다른 것으로 변경되면 domain 코드인 `Team`이 변경이 일어난다는 뜻입니다.

<img width="675" alt="스크린샷 2025-03-24 오후 7 07 06" src="https://github.com/user-attachments/assets/c6112200-8d37-458a-a54c-6d3423bc9894" />

praisebak replied:
> 위 부분은 지금은 그대로 사용하고 있지만 요구사항이 영어로 바뀌면 뷰에서 조건문으로 처리할것같습니다!
위와같이 구현한다고 해도 Team의 도메인 로직이 변경되어야할까요?

만약 영어로 바뀐다고하면 다음과 같이 바꾸면 될 것 같은데
어떻게 생각하시나요 오리??
```
if(team == Team.Blue){
  System.out.println(영어 출력);
}
```

jinyoungchoi95 replied:
영어가 되었을 때 새로 기능을 만드는게 아니라 "홍" "청"부터 옮기는것이 요구사항을 지키는 것 아닐까요? 명확하게 화면의 출력요소인데 domain이 가져야할 이유는 없는 것 같아요.

jinyoungchoi95 commented on src/main/java/piece/Team.java:
```diff
@@ -0,0 +1,28 @@
+
+    private final String type;
+
+    public static Team from(String s) {
```
`Team`의 테스트는 필요 없을까요? `Team`을 사용하는 클래스의 테스트로 커버가 된다고 한다면 그 클래스가 제거되면 `Team`은 무엇을 믿고 사용할 수 있을까요?

praisebak replied:
그렇군요 앞으로는 테스트를 더 작성하겠습니다!

praisebak replied:
Team에 대한 테스트에 대해 생각하니 다음과 같은 생각이 들어서 오리와 얘기해보고 싶습니다!

제가 이전에 드렸던 Team에 대한 테스트가 타 클래스에 의해 커버가 되기때문에 테스트를 작성하지않아도된다는 잘못된 결론 도출이었떤 것 같습니다.
오리가 말씀해주신대로 Team을 사용하던 타 클래스의 테스트나 클래스 자체가 삭제되면 Team은 커버가 안되게 되는 것이기 떄문입니다.

그것과 별개로 저는 Team에 대한 테스트가 필요없다고 생각합니다.
그 이유는 Team 그 자체의 도메인의 요구사항에서 중요도가 떨어지기 때문입니다.
Team 그 자체는 팀이라는 개념을 나타내는것이고 장기에서 중요한것은 Team을 기반으로 피스들을 관리하는게 중요합니다.
즉 그 Team을 사용하는 관리 객체의 테스트가 중요하고 그에 반해 Team은 작성하지않아도 된다는게 제 입장입니다

저는 LocalDateTime같은 객체를 사용할때, 해당 객체에 대한 테스트를 작성하지않습니다.
그 이유는 이미 검증이 됐기때문에 추가로 테스트를 작성하지 않아도 된다는 이유였습니다.
여기서 Team도 너무 간단한 객체이기때문에 추가 테스트가 필요없다고 보는 입장입니다.

jinyoungchoi95 replied:
간단하다 간단하지 않다는 주관적인 의견인 것 같아요. 

```java
    public Team opposite() {
        if (this == RED) {
            return BLUE;
        }
        if (this == BLUE) {
            return RED;
        }
        throw new IllegalStateException(NOT_SUPPORTED_TEAM);
    }
```
코드의 분기가 들어간 것만 해도 이미 로직입니다. 짧고 간단한 코드에 대한 기준보다는 테스트의 중요도 자체에 대해서만 바라보세요.

테스트가 짜여있지 않다면 `RED`의 반대는 `BLUE`라는 것을 누가 보장해줄 수 있나요? 이전 단계에서 TDD를 하셨다면 여기서도 TDD가 된다면 당연히 테스트가 작성되어있어야하지 않을까요? 리팩터링 과정에서 코드의 안정성을 보장해주는건 테스트밖에 없으니까요.

praisebak replied:
그렇군요 안정성 측면에서 테스트를 고려해야하는데 기능의 크고 작음으로 테스트를 생각했던것같습니다.

jinyoungchoi95 commented on src/main/java/piece/MoveRule.java:
```diff
@@ -0,0 +1,48 @@
+
+public class MoveRule {
+
+    private final MoveStrategy moveStrategy;
+    private final PieceType pieceType;
```
`MoveStrategy`와 `PieceType`가 따로 존재해야하는 이유가 무엇인가요? 네이밍만 보아도 서로 같은 존재인데 둘의 존재가 따로 있을 필요가 있는지 잘 모르겠습니다.

포 전략을 가지고 궁 타입을 가지는 `MoveRule`이 생성될 수도 있겠네요.

praisebak replied:
> MoveStrategy와 PieceType가 따로 존재해야하는 이유가 무엇인가요? 네이밍만 보아도 서로 같은 존재인데 둘의 존재가 따로 있을 필요가 있는지 잘 모르겠습니다.

기물 타입을 확인할때 PieceType를 사용하지않고 MoveStrategy만으로 instanceof를 사용해서 구분하는 것도 가능합니다
그런데 행동을 가지고 동일한지 비교하는것과 이름(PieceType)으로 비교하는것중 이름으로 비교하는것이 더 명확하고
나중에 행위와 별개로 이름을 분리해서 처리해야하는 부분에서 유리하지않겠느냐 생각하였습니다
MoveStrategy가 기물마다 다르니까 이것으로 구분할 수 있다는것은 도메인의 지식을 다 알고있을때 가능한 발상이라고 보고 명확성이 떨어진다고 생각하였습니다.

즉 저는 동기화보다는 명확성을 더 높게 보고 위와 같이 구현하였습니다. 
오리가 말씀해주신말씀대로 
> 포 전략을 가지고 궁 타입을 가지는 MoveRule이 생성될 수도 있겠네요.

저는 이부분을 동기화, 사용자가 잘 해야하는 영역이라고 생각하였습니다.
이 애플리케이션에서는 피스를 만드는 팩토리가 그 사용자가 될 것입니다.

아쉬운것은,장기가 잘 돌아가기 위해서는 현재 팩토리에서 피스들이 제각각 잘 들어갔는지가 검증이 되어야하는데,
이를 확인하는 테스트가 빈약한것같아 이부분 추가해보겠습니다

jinyoungchoi95 replied:
> 저는 이부분을 동기화, 사용자가 잘 해야하는 영역이라고 생각하였습니다.

사용자는 어떻게든 잘못 객체를 생성한다라는 관점으로 보시면 좋을 것 같네요. 잘 사용하는 것을 기대한다는 것으로 결론이 나버리면 테스트도 무의미해보여요. 테스트가 없어도 사용자가 잘 사용하면 되니까요.

instanceof를 사용하라는 의미에서 남긴 코멘트는 아니었고, `MoveBehavior`의 각 구현체가 `PieceType`을 가지고 있으면 해결되는 문제가 아닐까요? 지금 구조에서는 말씀드린 비정상 객체가 생성되는 것을 막기가 많이 힘들어보이네요.

praisebak replied:
MoveRule이 잘못 생성될 수 있다는 점에서 방어적 설계가 부족했던 것 같습니다.
- 대신 PieceType을 분리하는게 명확하다고 생각했습니다.

그러나 MoveBehavior에서 가지고 있어도 명확하면서 MoveRule이 잘못 생성되는 문제를 해결할 수 있으니 MoveBehavior이 가지고 있는 방식으로 수정하도록 하겠습니다!

praisebak replied:
[refactor: 불필요한 MoveRule 객체 분해하여 PieceType Behavior가 가지고 있도록 수정](https://github.com/woowacourse/java-janggi/pull/42/commits/0fe135d2eca6b1216e59487834c7d382df6c2cfe)

jinyoungchoi95 commented on src/main/java/strategy/MoveStrategy.java:
```diff
@@ -0,0 +1,49 @@
+
+    Route getLegalRoute(Position startPosition, Position endPosition, Team team);
+
+    default Route getLegalRoute(Position startPosition, Position endPosition, List<Route> canMoveDirections) {
```
default는 왜 나온 기능이라고 생각하시나요? 어떤 때에 활용하는 것이 적절할까요? 그 활용법이 객체지향이 맞을까요?

praisebak replied:
새로운 기능이 추가되었을때의 상속 방식의 한계, 또한 중복코드를 제거할 수 있는 이점을 위해서 나온것이라고 생각합니다.

기존 인터페이스에서는 모든 implemenet를 할때마다 아무리 중복코드가 생겨도 어쩔수없이 구현해야하는 문제가 있었습니다.
그렇다고 abstract class를 사용하자니 isA가 설계상에서 깨지지 않는다면 괜찮겠지만, 그렇지 않는다면 하위 클래스들을 모두 수정해줘야하는 문제가 발생합니다.
java에서는 collection을 도입하려고하니 isA가 깨져서 자바측에서도 한계를 인정하고 interface에 default를 도입한것으로 알고있습니다.

abstract class와 interface의 차이점은 인스턴스 변수를 가지고 있느냐, 다중 상속이 가능하냐 인것같습니다
저는 이번에 interface를 행위를 의미하는 설계로써 사용을 하였고,
그 과정에서 중복코드를 없애기 위해 default를 사용하였습니다.
또한 행위를 구현하기 위해 각 구현체들이 필드가 달라야했고 이런 점들을 고려했을때 interface가 더 낫다고 생각하였습니다. 
그리고 다중 상속이 가능한것도 이점이라고 보았습니다.

default 메서드를 사용하면 상속의 단점인 부모 클래스의 의존되는게 아니냐하는 문제는 
현재 기물들이 isA이고 뭔가 추가로 기물이 움직이는 방식이 바뀌는 것이 상상이 가지않았고 
바뀌지않는 설계라 보고 default라 두었습니다.

특히 움직이는 로직의 구현 방식(필요로하는 경로를 받아 이동가능한지 확인하는 방식)은 바뀌지 않을것이라 생각하여 default 메서드를 도입한것입니다

jinyoungchoi95 replied:
근본적으로 interface의 default가 나온 이유는 중복코드를 제거하기 위함이 아닙니다. 기능확장을 위한 것이고 중복코드 제거는 그로 인해 보이는 부수적인 효과일 뿐이에요.

그럼 지금 이 default가 기능 확장을 요구되고 있냐를 고민해보면 기능 확장용으로 사용된 것이 맞나요? 지금 개발되고 있는 것에 하위호환성을 지킬 무언가가 따로 있어보이지도 않아요. 즉, 초기개발인데 하위호환성을 위해 default를 사용했다라는 말이 맞지 않아요.

고민의 포인트를 중복제거에 초첨을 맞추지마시고 default가 나온 의미대로 쓰고 있냐에 초점을 맞추어보세요. 투다가 이야기한 중복제거는 abstract class를 구현해야하는 것 아닐까요?

praisebak replied:
자바에서 default 메서드를 만든 의도는 이해했었지만,
기능적인 면만 봤을때 인터페이스가 다중 상속도 되기 때문에 이젠 거의 추상 클래스의 상위호환이아닌가? 생각을 하였고
자바에서 사용하는 방식이 아닌 상위호환이니까 마냥 더 좋겠지하면서 썼던것같습니다.
위 방식은 자바에서 권장하는 방식으로 사용하고 있던 기존 개발자들과 협업을 하게 하면 제 인터페이스를 보고 당황할 것 같습니다.
이번에는 추상클래스를 사용하는것이 의도에 맞는 것 같아 다음과 같이 수정하였습니다   

[refactor: 인터페이스 default 출현의 의도와 맞지 않는 코드 -> 추상 클래스 사용하도록 수정](https://github.com/woowacourse/java-janggi/pull/42/commits/332ca0b31c317a1a0be32b5cd1235a37d600785f)

jinyoungchoi95 commented on src/main/java/strategy/JolMoveStrategy.java:
```diff
@@ -0,0 +1,44 @@
+                        List.of(new Position(0, 1))
+                ),
+                new Route(
+                        List.of(new Position(0, -1))
```
위치라는 것은 9x9판에서의 위치를 뜻하는 것으로 보이는데 -1이 있는 것이 안와닿네요. `Position`만 놓고 보았을 때 column이 음수일 수 있나요?

praisebak replied:
구현하면서도 헷갈려서 추가해야한다고 생각했었는데,
결국 시간 관계상 다음 리팩토링으로 미뤘던 사항입니다.

Position과 Direction을 구분짓는 방식으로 수정하겠습니다!

praisebak replied:
[refactor: 장기판상 위치와 방향 객체 분리](https://github.com/woowacourse/java-janggi/pull/42/commits/094441bdf124d0d84900f5eb024a5176e766005a)

jinyoungchoi95 commented on src/main/java/strategy/JolMoveStrategy.java:
```diff
@@ -0,0 +1,44 @@
+
+public class JolMoveStrategy implements MoveStrategy {
+
+    private static final Map<Team, List<Route>> teamCanMoveDirection;
```
`canMoveDirections`이 어디는 static이고 어디는 static이 아닌데 차이가 있을까요?

praisebak replied:
아 이부분은 모두 static으로 해야했는데 제 실수로 static을 누락한것같습니다!

praisebak replied:
[refactor: static 누락 수정](https://github.com/woowacourse/java-janggi/pull/42/commits/6f876eec62619831f71d5e69c30d7112f2c78e0f)

jinyoungchoi95 reviewed (changes requested):
안녕하세요 투다 😄

장기 미션의 리뷰어를 맡은 오리입니다.

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

praisebak committed:
refactor: 장기 초기화 파일방식에서 초기 기물을 상태로서 가지고 있게 수정

praisebak committed:
refactor: 장기 초기화 파일방식에서 enum 방식으로 변경

praisebak committed:
refactor: static 누락 수정

praisebak committed:
refactor: static 누락 수정

praisebak committed:
refactor: 중복된 움직임 예외 처리 부분 가독성 향상

praisebak committed:
refactor: 중복된 움직임 예외 처리 부분 가독성 향상

praisebak committed:
refactor: 필요하지않은 기물 이동방식 팩토리 제거

praisebak committed:
refactor: 필요하지않은 기물 이동방식 팩토리 제거

praisebak committed:
refactor: 사용하지 않는 코드 제거 및 Position 불필요한 static 사용 제거

praisebak committed:
refactor: 사용하지 않는 코드 제거 및 Position 불필요한 static 사용 제거

praisebak committed:
refactor: 팩토리 명명 수정

praisebak committed:
refactor: 팩토리 명명 수정

praisebak committed:
test: 장기 기물 생성 테스트 추가

praisebak committed:
refactor: 사용하지 않는 클래스 파일 제거

praisebak committed:
refactor: 팀단위로 피스 관리와 포지션으로 기물들 관리하는 책임을 Board,TeamPieces로 분리

praisebak committed:
refactor: 명명 수정

Board라는 이름으로는 포지션단위로 장기판을 관리가 예상되지않음 -> PositionPieces
TeamPieces -> PlayerPieces

praisebak committed:
test: 팀 단위 기물 관리 테스트 추가

praisebak committed:
test: 포지션별 기물 관리 테스트 추가

praisebak committed:
refactor: 장기판상 위치와 방향 객체 분리

praisebak committed:
feat: 체크메이트 기능 구현

praisebak committed:
feat: 상차림 기능 구현

jinyoungchoi95 commented on src/main/java/move/MoveBehavior.java:
```diff
@@ -0,0 +1,56 @@
+import piece.Route;
+import piece.Team;
+
+public interface MoveBehavior {
+
+    @FunctionalInterface
+    interface MoveBehaviorThrowingPredicate {
+        boolean test();
+    }
+
+    default void throwInvalidMoveBehaviorByCondition(MoveBehaviorThrowingPredicate throwCondition) {
+        if (throwCondition.test()) {
+            throw new InvalidMovePosition();
+        }
```
복잡도만 가득한 코드인 것 같습니다. 인터페이스는 인터페이스답게 설계해보세요. 눈에 깔끔하게 코드를 작성하는 것보다 각각의 역할에 맞게 설계해보는 것이 1레벨의 목적입니다.

인터페이스는 기능 명세서인데 지금의 `MoveBehavior`는 그냥 구현체만 담고있는 클래스처럼 보이네요. 추상화가 잘 된 것이 맞나요?

praisebak replied:
인터페이스의 의도대로 하지 않았던 것 같습니다.
따라서 추상 클래스를 사용하도록 하여서 수정해보았습니다

jinyoungchoi95 commented on src/main/java/move/JolMoveBehavior.java:
```diff
@@ -0,0 +1,44 @@
+
+    static {
+        final List<Directions> blueCanMoveDirections = List.of(
+                new Directions(
+                        List.of(Direction.RIGHT)
+                ),
+                new Directions(
+                        List.of(Direction.LEFT)
+                ),
+                new Directions(
+                        List.of(Direction.DOWN)
+                )
```
```suggestion
                new Directions(List.of(Direction.RIGHT)),
                new Directions(List.of(Direction.LEFT)),
                new Directions(List.of(Direction.DOWN))
```
불필요한 공백은 제거되는 것이 좋겠네요

jinyoungchoi95 commented on src/main/java/move/FoMoveBehavior.java:
```diff
@@ -0,0 +1,79 @@
+
+    private void validateFoMove(Position destination, Team moveTeam, int onRoutePiecesSize, Piece firstPiece,
+                                Piece lastPiece) {
+        throwInvalidMoveBehaviorByCondition(() -> !(onRoutePiecesSize == 1 || onRoutePiecesSize == 2));
+        throwInvalidMoveBehaviorByCondition(() -> onRoutePiecesSize == 1 && firstPiece.isSamePosition(destination));
+        throwInvalidMoveBehaviorByCondition(() -> onRoutePiecesSize == 2 && !lastPiece.isSamePosition(destination));
+        throwInvalidMoveBehaviorByCondition(
+                () -> lastPiece.isSamePosition(destination) && lastPiece.isSameTeam(moveTeam));
```
처음 코드를 보는 사람은 `validateFoMove`를 볼때 59~63까지 모든 코드를 컨텍스트를 머릿속에 담아두면서 코드를 읽어야하는데요. 현재 구조라면 `onRoutePiecesSize가 1인 경우`, `onRoutePiecesSize가 2인 경우` 등으로 if 분기로 컨텍스트를 끊어주는 것이 좋겠네요. 굳이 각 validation을 모두 읽어야할 필요는 없으니까요.

praisebak replied:
[refactor: 컨텍스트를 고려한 validation으로 수정](https://github.com/woowacourse/java-janggi/pull/42/commits/fd5e05e64333d93c6d5d892860c3a4e44e73f9ba)

위와 같이 수정하였습니다!

jinyoungchoi95 commented on src/main/java/move/Direction.java:
```diff
@@ -0,0 +1,29 @@
+    DOWN_LEFT(-1, -1),
+    DOWN_RIGHT(-1, 1);
+
+    final int y;
+
+    final int x;
```
```suggestion
    private final int y;

    private final int x;
```

jinyoungchoi95 commented on src/main/java/piece/PositionPieces.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+import java.util.Map;
+
+public class PositionPieces {
+
+    private final Pieces allPieces;
```
`Pieces`를 한번 더 래핑한 클래스가 필요한 이유가 무엇일까요? 불필요한 래핑클래스로 인해서 class depth가 늘어 복잡도가 올라간 것 같네요.

jinyoungchoi95 commented on src/main/java/piece/PlayerPieces.java:
```diff
@@ -0,0 +1,53 @@
+    private static final String INVALID_PLAYER_SIZE = "장기는 2명이서 할 수 있습니다";
+    private static final int PLAYER_SIZE = 2;
+
+    private final Map<Team, Pieces> teamBoard;
```
초기 설계에서 이것저것 들어오다보면 어디서부터인가 꼬여서 중복되는 의미가 많이 담기게 되는데요. 현재 기준으로 말씀드리면

1. `PlayerPieces`가 `Map<Team, Pieces>`를 가지고 있고
2. `Pieces`가 `List<Piece>`를 가지고 있고
3. `Piece`가 아래를 가지고 있습니다.
```java
    private final MoveRule moveRule;
    private final Team team;
    private Position position;
```

`Team`이라는 정보가 `PlayerPieces`에도 있고 `Piece`에 있는데 이러면 팀을 구분짓는 책임이 여러군데 혼합되어있는 것이나 마찬가지인거죠.

여러 객체가 똑같은 정보(책임)을 다른 곳에서 가지는 것이 아닌 하나의 객체가 본인의 책임을 가질 수 있도록 설계해보면 좋을 것 같네요.

praisebak replied:
몇가지 질문이 있어요!
> 여러 객체가 똑같은 정보(책임)을 다른 곳에서 가지는 것이 아닌 하나의 객체가 본인의 책임을 가질 수 있도록 설계해보면 좋을 것 같네요.

제가 이번에 위와 같이 Team을 단위로 구분지은 이유는 
상태상에서 봤을때는 Pieces에서 piece에게 team을 물어보는 식으로 구현할 수 있지만,
플레이어가 2명 나뉘어서 플레이하는 장기의 특성상, 팀 단위로 구분짓는것이 좀 더 책임이 이해하기 쉽고 명확하다고 느꼈기 떄문입니다.

Places는 팀에 대해서는 관계없이 기물들을 관리하는 메시지를,
PlayerPieces는 팀에 대해서 해당하는 Pieces에 대해서 기물들을 관리하는 메시지를 관리하는 객체로 나누고 싶었습니다.

만약 Pieces에서 팀에 해당하는 정보들을 추출하는 메서드들까지 포함하게 되면 너무 Places에서 책임이 많아지는게 아닌가 생각이 들었습니다.

위와 같이 명확도를 위해 중복된 관리할 부분이 생기는 것을 저는 트레이드오프의 영역이라고 생각했었는데
이부분 오리의 의견이 궁금합니다!

jinyoungchoi95 replied:
저는 트레이드오프의 영역이라 한다면 서로 다른 두 선택지가 모두 크리티컬한 단점이 있어야한다고 생각합니다.

지금의 문제는 `Team`이라는 정보가 중복관리되고 있다는 것이고 그로 인해서 `Team`이 서로 다른 값이 잘못 저장될 수 있다는 문제가 있는 것이에요. 반대의 케이스는 말씀하신 단순히 책임이 많아지는 것 그이상은 없는 것이구요.

우선 `Piece`가 `Team`을 가지고 있는 관점에서 이야기를 하면, Map없이 `List<Piece>`만 있어도 되는 것 아닐까요? 움직이려는 위치의 `Piece`도 `Team`을 가지고 있을 테니 그것과 단순 비교만 하면 되는 것이니까요.
1. `Piece`를 움직인다.
    a. 현재 턴의 `Team`이 아닌 `Piece`를 움직이려하면 예외를 던진다.
2. 움직인 위치의 `Piece`가 어떤 팀인지에 따라 결정한다.
를 하면 `PlayerPieces`가 딱히 key로 Team을 가져야하는 이유가 없는 것 같아요.

설계를 하실 때에는 책임을 분리하고 객체 설계를 잘 하는 것도 중요하지만 그게 과한 설계이느냐를 계속 고민해보시면 좋겠어요. 제가 리뷰드리는 과함의 기준은 꼭 이게 있어야만 되느냐입니다. 출석때 있었던 `구현 요구사항을 만족하는 최소한의 설계를 한다.`를 계속 인지해보시는 것도 도움이 될 것 같아요.

jinyoungchoi95 commented on src/main/java/piece/Position.java:
```diff
@@ -0,0 +1,66 @@
+package piece;
```
`piece`에 있는 패키지 클래스들이 많은 것 같아요. `piece` 내부에서도 비슷한 성격을 가진 클래스들은 또다른 하위 패키지로 묶어보면 좋겠네요.

jinyoungchoi95 commented on src/main/java/piece/Route.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Route {
+
+    private final List<Position> positions;
```
`Route`는 일급컬랙션으로 가지는 장점이 무엇이 있을까요? 불변이 장점인 것이라면, 사용처들에서 불변 List를 반환해주면 되는 것이 아닐까요? (필요한 래핑 클래스인지 고민해보세요.)

praisebak replied:
불필요한 객체인것같습니다.
사용하면서도 크게 책임이없는데 억지로 만들어 사용하는 느낌이 강했던것같습니다.
삭제하도록 하였습니다!

jinyoungchoi95 commented on src/test/java/PositionTest.java:
```diff
@@ -0,0 +1,32 @@
+class PositionTest {
+    @Test
+    void 포지션은_숫자를_가진다() {
+        // given
+
+        // when
```
사용하지 않는 주석은 제거해주세요. 17번줄로도 표현이 가능해서 bdd의 포맷을 완전히 지킬 필요는 없는 것 같네요.

praisebak replied:
tdd 포맷을 이번에 처음 적용해보았는데, 저에게는 별로 맞지않는것같아 삭제하였습니다!

jinyoungchoi95 replied:
해당 주석은 tdd 포맷은 아니고 bdd 포맷이에요.

https://en.wikipedia.org/wiki/Behavior-driven_development

jinyoungchoi95 commented on src/test/java/move/ChaMoveBehaviorTest.java:
```diff
@@ -0,0 +1,62 @@
+import piece.Pieces;
+import piece.Position;
+import piece.Route;
+import piece.Team;
+
+public class ChaMoveBehaviorTest {
```
```suggestion
import piece.Team;

class ChaMoveBehaviorTest {
```
class 선언이 public인 것과 섞여있는 것 같네요.

jinyoungchoi95 commented on src/test/java/move/FoMoveBehaviorTest.java:
```diff
@@ -0,0 +1,90 @@
+
+        // then
+        List<Position> positions = route.positions();
+        Assertions.assertThat(positions.size()).isEqualTo(5);
```
```suggestion
        Assertions.assertThat(positions).hasSize(5);
```
`assertAll`이라는 키워드에 대해서 학습해볼까요?

jinyoungchoi95 commented on src/test/java/PositionTest.java:
패키지가 안맞는 것 같네요.

jinyoungchoi95 commented on src/main/java/piece/Position.java:
```diff
@@ -0,0 +1,66 @@
+        return this.column == position.column;
+    }
+
+    private boolean isAbsoluteBigger(Position position) {
+        return (this.row + this.column) - (position.row + position.column) > 0;
+    }
+
+    public Position getSmallerPosition(Position otherPosition) {
+        if (this.isAbsoluteBigger(otherPosition)) {
+            return otherPosition;
+        }
+        return this;
+    }
+
+    public Position getBiggerPosition(Position otherPosition) {
+        if (this.isAbsoluteBigger(otherPosition)) {
+            return this;
+        }
+        return otherPosition;
+    }
```
메서드 정렬기준이 있을까요?

praisebak replied:
음 실행되는 순서에 맞춰서나 읽기 좋은 순서대로 정렬해야하지만
잘 지켜지지 않은것 같아요.
위에 isAbsoluteBigger가 있으니 다음도 getBiggerPosition가 오면 좋을것같습니다
앞으로는 신경써보도록하겠습니다!

jinyoungchoi95 commented on src/main/java/piece/Pieces.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.List;
+import java.util.Map;
+
+public class Pieces {
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.

전반적으로 depth에 대해서 신경써주면 좋겠네요.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 투다 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

praisebak committed:
refactor: 인터페이스 default 출현의 의도와 맞지 않는 코드 -> 추상 클래스 사용하도록 수정

praisebak committed:
refactor: 움직임 예외 exception 패키지 -> move 패키지로 이동

praisebak committed:
refactor: 불필요한 MoveRule 객체 분해하여 PieceType Behavior가 가지고 있도록 수정

praisebak committed:
refactor: 컨텍스트를 고려한 validation으로 수정

praisebak committed:
refactor: 불필요한 래핑 삭제

너무 책임이 작은 경우 래핑해도 이점이없고 클래스 depth만 깊어질 수 있다

praisebak committed:
chore: 패키지 구조 세밀하게 분리

praisebak committed:
refactor: route 제거

praisebak committed:
refactor: 테스트 클래스에서 public 제거

praisebak committed:
refactor: assertAll 적용

praisebak committed:
refactor: hasSize 적용

praisebak committed:
fix: 테스트 위한 상수 깨진 문제 수정

praisebak committed:
refactor: intent 2 -> intent 1 왠만하면 준수하도록 수정 (2중 포문 제외)

praisebak committed:
test: 팀 테스트 추가

praisebak commented:
이번 피드백을 통해 다음을 배웠습니다.
- 개발자들간 많이 알려진 일반적인 통념을 꺨려면 그에 상응하는 이유가 없으면 깨면 안된다(암묵적 약속이라고 볼 수 있음)
 - 스스로 쓰면서 불편했다면 그것이 필요한지 의심해보자
 - 방어적 설계와 명확성 저울질 - 명확성에서 아주약간 이점을 얻기위해 방어적 설계를 포기했었음 (저울질을 잘하자)
 - 유틸,아키텍쳐,디자인패턴 같은 개념들은 객체지향에 상관없다. 나중에 필요해지면 사용하면 되니 잠시 잊어두자.
 - 관리요소 증가 vs 명확성은 트레이드오프인가?
 - (클린코드) - 가독성 컨텍스트, 이름상 적합한 상태,로직인지
 - 테스트는 안정성. 높아서 나쁠것이 없다

jinyoungchoi95 commented on src/main/java/TurnResult.java:
```diff
@@ -0,0 +1,5 @@
+import java.util.Optional;
+import piece.Team;
+
+public record TurnResult(int nextTurn, Optional<Team> loseTeam) {
```
`Optional`은 반환타입으로만 사용하는 것이 맞습니다. 변수로 들고 있으면 어떤 문제가 발생할까요?

https://mangkyu.tistory.com/203

jinyoungchoi95 commented on src/main/java/piece/PlayerPieces.java:
```diff
@@ -0,0 +1,51 @@
+        return bluePieces.add(redPieces);
+    }
+
+    public Optional<Team> kingDeadTeam() {
```
위와는 별개로 해당 메서드에서 `Optional`을 쓰는 것이 맞는지도 고민해보면 좋겠네요.

"있다, 없다"의 상황처리 책임을 외부로 넘기는 것 아닐까요?

jinyoungchoi95 commented on src/main/java/piece/PlayerPieces.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    public Pieces allPieces() {
+        Pieces bluePieces = teamBoard.get(Team.BLUE);
+        Pieces redPieces = teamBoard.get(Team.RED);
```
현재 구조는 2팀을 제한하고 있고, `allPieces`의 기능 상 BLUE와 RED만이 모든 피스인 것으로 표현이 되고 있어요.

그럼
```java
    private final Pieces redPieces;
    private final Pieces bluePieces;
```
로 표현하는 것과 무엇이 다를까요? `Collection`의 장점을 활용하고 있는 것이 맞나요?

jinyoungchoi95 reviewed (approved):
안녕하세요 투다 😄

답변과 몇가지 코멘트 추가로 남겨두었으니 다음 단계 진행하면서 함께 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
